catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/history/Historian.java

```diff
@@ -0,0 +1,502 @@
+
+    private final Map<String, List<Callable<Boolean>>> staged = new HashMap<>();
+
+    /**
+     * Versions {@link Metacard}s from the given {@link CreateResponse}.
+     *
+     * @param createResponse Response to create versioned metacards from
+     * @return The original {@link CreateResponse}
+     * @throws IngestException
+     */
+    public CreateResponse versionCreate(CreateResponse createResponse) throws IngestException {
+        if (doSkip(createResponse)) {
+            return createResponse;
+        }
+        setSkipFlag(createResponse);
+
+        List<Metacard> versionedMetacards = createResponse.getCreatedMetacards()
+                .stream()
+                .filter(MetacardVersion::isNotVersion)
+                .map(m -> new MetacardVersion(m,
+                        MetacardVersion.Action.CREATED,
+                        SecurityUtils.getSubject()))
+                .collect(Collectors.toList());
+
+        if (!versionedMetacards.isEmpty()) {
+            getSystemSubject().execute(() -> catalogProvider().create(new CreateRequestImpl(
+                    versionedMetacards)));
+        }
+
+        return createResponse;
+    }
+
+    /**
+     * Version updated metacards based off of the successfully created metacards found in the
+     * {@link UpdateResponse}
+     *
+     * @param updateResponse Versioned metacards created from any created metacards in this
+     * @return The original UpdateResponse
+     * @throws SourceUnavailableException
+     * @throws IngestException
+     */
+    public UpdateResponse versionUpdate(UpdateResponse updateResponse)
+            throws SourceUnavailableException, IngestException {
+        if (doSkip(updateResponse)) {
+            return updateResponse;
+        }
+        setSkipFlag(updateResponse);
+
+        List<Metacard> inputMetacards = updateResponse.getUpdatedMetacards()
+                .stream()
+                .map(Update::getNewMetacard)
+                .collect(Collectors.toList());
+
+        CreateResponse createdHistory = getVersionedMetacards(inputMetacards,
+                MetacardVersion.Action.UPDATED);
+        if (createdHistory == null) {
+            return updateResponse;
+        }
+
+        if (updateResponse.getProperties() != null) {
+            updateResponse.getProperties()
+                    .put(HISTORY_METACARDS_PROPERTY,
+                            new ArrayList<>(createdHistory.getCreatedMetacards()));
+        }
+
+        return updateResponse;
+    }
+
+    /**
+     * Versions created {@link Metacard}s and {@link ContentItem}s.
+     *
+     * @param createStorageRequest Versions this requests items
+     * @return The transaction key for using with {@link Historian#commit(String)} and
+     * {@link Historian#rollback(String)}
+     * @throws IngestException
+     * @throws StorageException
+     */
+    public Optional<String> versionStorageCreate(CreateStorageRequest createStorageRequest)
```

Comment:
This is fairly long method. Could you break it down into smaller steps to make it easier to understand?